### ISSUE TITLE
Make a homogeneous code

### ISSUE THREAD

**Issue by saeid-h** [cogaplex-bts/bts#30]
I am interested in your work. I wonder if it is possible to eliminate the c++ and cuda part to make the code more simple and readable. If you use TensorFlow it backpropagates the gradient automatically. It also helps to use your code in future work and make some changes easily. Here is my suggestion for the elimination of cuda part in BTS v1. I am not sure about my transformation because some parts are not described in the paper. Please consider it: [https://github.com/saeid-h/bts-fully-tf](https://github.com/saeid-h/bts-fully-tf)

**Comment by cogaplex-bts** [comment]
Really thanks for your contribution. We had some reasons for we implementing the LPG layers using C++ API. The main concern was to use analytic gradient not the numeric gradient computed automatically by TF. However, we agree with you that current implementation prevents users easy leading and adopting for their own application. Currently we are working with pytorch implementation of BTS, and I believe your brilliant implementation of the LPG layer using only python functions will be adopted to our codes. Thanks! =)

**Comment by saeid-h** [comment]
Thank you for the great job you've done. I'm looking forward for updates.

**Comment by huangxGo** [comment]
> @saeid-h Really thanks for your contribution. We had some reasons for we implementing the LPG layers using C++ API. The main concern was to use analytic gradient not the numeric gradient computed automatically by TF. However, we agree with you that current implementation prevents users easy leading and adopting for their own application. Currently we are working with pytorch implementation of BTS, and I believe your brilliant implementation of the LPG layer using only python functions will be adopted to our codes. Thanks! =)

Hello! Thanks for the awesome work!
Do you plan to release the pytorch version of the BTS?
